### DISCUSSION TITLE
Load Avalonia ResourceDictionary build in an external dll

### DISCUSSION THREAD

**Discussion by lebhero** [AvaloniaUI/Avalonia#16954] [General]
I created a c# class project where i added an AvaloniaResourceDictionary called Strings.fr.axaml which contains some system strings:

```
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
                    xmlns:system="clr-namespace:System;assembly=System.Runtime"
                    >
    <system:String x:Key="Hello">Bonjour</system:String>
    <system:String x:Key="Goodbye">Au revoir</system:String>
</ResourceDictionary>
```

i also add the AvaloniaResource as an itemgroup:

```
<ItemGroup>
  <AvaloniaResource Include="Strings.fr.axaml" />
</ItemGroup>

<ItemGroup>
  <PackageReference Include="Avalonia" Version="11.1.3" />
</ItemGroup>
```
  
  I can now build the project which will generate a dll.

in the App.axaml.cs file i added the following function:

```
private void LoadLanguage(string languageDllPath)
{
    try
    {
        // Load the assembly dynamically
        var assembly = Assembly.LoadFrom(languageDllPath);
        

        var resourceNames = assembly.GetManifestResourceNames();
        foreach (var resourceName in resourceNames)
        {
            Debug.WriteLine(resourceName);
        }

        var types = assembly.GetTypes();
        string x = types[10].FullName; // this is only from testing     

        // Find the ResourceDictionary in the loaded assembly
        var resourceDictType = assembly.GetType(x);

        if (resourceDictType != null)
        {
            var resourceDictionary = (ResourceDictionary)Activator.CreateInstance(resourceDictType);
            // Add the ResourceDictionary to the application's resources
            Application.Current.Resources.MergedDictionaries.Add(resourceDictionary);
        }
    }
    catch (Exception ex)
    {
        // Handle errors, such as if the assembly or resources can't be found
        Console.WriteLine($"Failed to load language resources: {ex.Message}");
    }
}
```

![image](https://github.com/user-attachments/assets/46f0b4b1-6c4b-4447-ad03-1bb17d6ff28b)

In ILSpy i see this:

![image](https://github.com/user-attachments/assets/207c5c88-2865-4869-8f9b-bfa4c51c33ce)

Unfortunately i am unable to load the ResourceDictionary so that the binding in the textblock below does not work:

`<TextBlock Text="{DynamicResource Goodbye}" Background="Red"/>`

Can anyone please help on how to load the dll and fetch the resourcedictionary correctly?
Thanks in advance

**Comment by thevortexcloud** [comment]
I don't know how to solve this specific problem, but a possible work around is to just embed the resource dictionary itself and then load it using the `avares://` URI without messing with manually loading assemblies (assuming you have a compile time reference to the other assembly).

https://docs.avaloniaui.net/docs/basics/user-interface/assets

You can then reference it from within your normal XAML.

**Reply by lebhero** [reply]
Thank you for the suggestion... but the approach i am tending to do is to make sure i can support more languages in my application without having to build and release the complete application.

**Comment by rabbitism** [comment]
Option 1: you can define a class behind ResouceDictionary, then you can just construct it in code. Example https://github.com/irihitech/Semi.Avalonia/tree/main/src/Semi.Avalonia/Locale

Option 2: use AvaloniaRuntimeXamlLoader

**Comment by MarkoShenodaa** [comment]
Try add this to your App.axmal
<Application.Resources>
    <ResourceDictionary>
        <ResourceDictionary.MergedDictionaries>
            <ResourceInclude Source="avares://Your Dll Namespace/If any forlders add here/ResourceDictionary.axaml"/>
        </ResourceDictionary.MergedDictionaries>
    </ResourceDictionary>
</Application.Resources>
